### DISCUSSION TITLE
Download file by click in synchronous mode

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3802] [General]
Hello! I`m trying to automate scenario of downloading video from my own youtube channel:

When I click Скачать (Download) in browser the file save dialog appears. I found the solution to use here https://stackoverflow.com/questions/28565319/force-cefsharp-to-download-without-showing-dialog , but its asynchronous. What if I need to wait for download to be finished before getting to next algorithm step? So my canonical code should look like this:

Navigate();
ClickDownloadButton();
WaitForDownloadToFinish();
DoSmthWithDownloadedFile();

How could I organize this?

**Comment by amaitland** [comment]
Downloading is handled asynchronously, I personally would discourage trying to make it synchronous.

You can use the new fluent DownloadHandler implementation to simplify the process.

https://github.com/cefsharp/CefSharp/issues/3550#issuecomment-851832202

There is also some new method for downloading files given a url. See https://github.com/cefsharp/CefSharp/pull/3576/files#diff-b5d302e3b07bf84e99ffa745fe37affd00a70cae3e5dd66c51911e30540c06c1R342

**Reply by assopri** [reply]
Can I somehow stop execution of the high level algorithm to wait for the download handler to finish?

**Reply by amaitland** [reply]
Stop execution of what exactly?

**Reply by assopri** [reply]
Wait (2) after click 
![image](https://user-images.githubusercontent.com/73235464/132956148-994e069c-ad61-46c8-8978-4620089eae22.png)
to be sure that after WaitForDownloadToFinish returned, download is finished.

**Comment by ukandrewc** [comment]
You can use a DispatcherFrame or AutoResetEvent to wait for the download to complete.

You would initially need to wait for the OnBeforeDownload event with a timeout, because it may not be raised, if there is an error or the URL is invalid.

**Comment by amaitland** [comment]
> You can use a DispatcherFrame

DispatcherFrame is intended to be used in WPF. WinForms doesn't use a dispatcher. 

@assopri What exactly is your requirement to run your methods sequentially?  Something like the following will be much easier to implement and avoids a number of different threading issues.

```c#
using CefSharp.Fluent;

//Save all files to temp folder, update to whatever suites you.
var userTempPath = System.IO.Path.GetTempPath();

chromiumWebBrowser.DownloadHandler =
                    DownloadHandler.UseFolder(userTempPath,
                        (chromiumBrowser, browser, downloadItem, callback) =>
                        {
                            if(downloadItem.IsComplete)
                            {
                                DoSmthWithDownloadedFile();
                            }                         
                        });

```

If you can obtain the url you can avoid using a download handler all together and simply directly download the file https://github.com/cefsharp/CefSharp/pull/3576/files#diff-b5d302e3b07bf84e99ffa745fe37affd00a70cae3e5dd66c51911e30540c06c1R342

**Reply by ukandrewc** [reply]
> DispatcherFrame is intended to be used in WPF. WinForms doesn't use a dispatcher.

You can use it in Winforms as well.

**Reply by amaitland** [reply]
> You can use it in Winforms as well.

Just because you can use the method doesn't mean you should use it. There is no Dispatcher assigned to the WinForms UI thread by default unlike the WPF UI thread for which the dispatcher was initially implementation. 

Creating a dispatcher also creates a hidden window, allocates a number of other resources, which just aren't nessicary in WinForms. 

- https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/Dispatcher.cs,352
- https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/Dispatcher.cs,63
- https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/Dispatcher.cs,1934

**Reply by ukandrewc** [reply]
> Just because you can use the method doesn't mean you should use it.

Just because you won't, doesn't mean that others can't.

**Reply by amaitland** [reply]
I'd encourage anyone interested to take a few minutes to read through the code, step through it with the debugger would be even better. There is a lot going on under the hood for what is effectively manually pumping the message loop.

https://docs.microsoft.com/en-us/visualstudio/debugger/how-to-debug-dotnet-framework-source?view=vs-2019

If you absolutely have to pump the message loop in WinForms then [Application.DoEvents](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.doevents?view=net-5.0) will have minimal overhead comparatively speaking. 

Pumping the message loop in any context can be problematic and I'd encourage reading https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.doevents?redirectedfrom=MSDN&view=net-5.0#remarks

**Reply by assopri** [reply]
> ```cs
> DownloadHandler
> ```

THank you for help! I use DownloadHandler . Just one question is still not solved - how can I cancel showing the separate popup form, appearing after I click on download link?

**Reply by assopri** [reply]
I mean, what should I add here:
`cefBrowserWrapper.DownloadHandler = DownloadHandler.UseFolder("C:\\Users\\User\\Documents\\ViberDownloads",
                        (chromiumBrowser, browser, downloadItem, callback) =>
                        {
                            if (downloadItem.IsComplete)
                            {
                                downloadReadyEvent.Set();
                            }
                        });`
Because here I just set OnDownloadUpdateDelegate in this case (as 2nd parameter) instead of OnBeforeDownloadDelegate. And I cannot to anything like this:
![image](https://user-images.githubusercontent.com/73235464/133539396-703c3e9b-7fb1-4de1-a4bb-74d5a9fd86db.png)

**Reply by amaitland** [reply]
> how can I cancel showing the separate popup form, appearing after I click on download link?

You can close the popup after the download is complete by using something like the following:

```c#
chromiumWebBrowser.DownloadHandler =
                    DownloadHandler.UseFolder(userTempPath,
                        (chromiumBrowser, browser, downloadItem, callback) =>
                        {
                            if(downloadItem.IsComplete)
                            {
                                if(browser.IsPopup && !browser.HasDocument)
				{
				  browser.GetHost().CloseBrowser(true);
				}
                            }
                        });
```

> Because here I just set OnDownloadUpdateDelegate in this case (as 2nd parameter) instead of OnBeforeDownloadDelegate. And I cannot to anything like this:

Not sure what you are trying to do here exactly. If you are finished with the download handler instance just set the `DownloadHandler` to `null` and assign a different instance later.

```c#
browser.DownloadHandler = null;
```

If that's not what you mean then please provide more detail.

To make sure code more readable please use formatting as per https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by assopri** [reply]
> > how can I cancel showing the separate popup form, appearing after I click on download link?
> 
> You can close the popup after the download is complete by using something like the following:
> 
> ```cs
> chromiumWebBrowser.DownloadHandler =
>                     DownloadHandler.UseFolder(userTempPath,
>                         (chromiumBrowser, browser, downloadItem, callback) =>
>                         {
>                             if(downloadItem.IsComplete)
>                             {
>                                 if(browser.IsPopup && !browser.HasDocument)
> 				{
> 				  browser.GetHost().CloseBrowser(true);
> 				}
>                             }
>                         });
> ```
> 
> > Because here I just set OnDownloadUpdateDelegate in this case (as 2nd parameter) instead of OnBeforeDownloadDelegate. And I cannot to anything like this:
> 
> Not sure what you are trying to do here exactly. If you are finished with the download handler instance just set the `DownloadHandler` to `null` and assign a different instance later.

I mean, I want not to close the popup, but avoid showing it at all.

**Reply by amaitland** [reply]
> I mean, I want not to close the popup, but avoid showing it at all.

To stop showing it at all then you'll need to change the behaviour of the link that initiated the download. For example remove the `target="_blank"` from the `<a/>` tag will avoid creating a popup/tab. You can probably achieve this via javascript.

The default behaviour in `Chrome` is to open a tab and to remove it once the download has started. 

For `CEF` you can hide the popup once the download has started and close it once the download is complete. The `popup` will need to remain hidden until the download is complete, then it can be closed.

```c#
[DllImport("user32.dll")]
[return: MarshalAs(UnmanagedType.Bool)]
private static extern bool IsWindowVisible(IntPtr hWnd);

[DllImport("user32.dll")]
private static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);

var tempPath = System.IO.Path.GetTempPath();

browser.DownloadHandler =
	Fluent.DownloadHandler.UseFolder(tempPath,
		(chromiumBrowser, browser, downloadItem, callback) =>
		{
			if (downloadItem.IsComplete || downloadItem.IsCancelled)
			{
				if (browser.IsPopup && !browser.HasDocument)
				{
					browser.GetHost().CloseBrowser(true);
				}
			}
			//TODO: You may wish to customise this condition to better suite your
			//requirements. 
			else if(downloadItem.ReceivedBytes < 100)
			{
				var popupHwnd = browser.GetHost().GetWindowHandle();

				var visible = IsWindowVisible(popupHwnd);
				if(visible)
				{
					const int SW_HIDE = 0;
					ShowWindow(popupHwnd, SW_HIDE);
				}
			}
		});
```

**Reply by assopri** [reply]
Cool! That was what I was really looking for!

**Reply by assopri** [reply]
Now I really feel happy and very excited! Thank you!!!

**Comment by ukandrewc** [comment]
Application.DoEvents() is intended to allow UI refresh. Using it in a loop to wait for another task, you also need to sleep the thread otherwise it will consume 100% processor time. That is why I suggested DispatcherFrame or ResetEvent.

I agree that downloading async is best, but that's not the question :-)
